crates/subspace-archiving/src/piece_reconstructor.rs

```diff
@@ -0,0 +1,176 @@
+    /// Erasure coding data structure
+    reed_solomon: ReedSolomon,
+    /// KZG instance
+    kzg: Kzg,
+}
+
+impl PiecesReconstructor {
+    fn shards_count(segment_size: u32, record_size: u32) -> u32 {
+        segment_size / record_size
+    }
+
+    fn data_shards(&self) -> u32 {
+        Self::shards_count(self.segment_size, self.record_size)
+    }
+
+    fn parity_shards(&self) -> u32 {
+        self.data_shards()
+    }
+
+    pub fn new(
+        record_size: u32,
+        segment_size: u32,
+        kzg: Kzg,
+    ) -> Result<Self, ReconstructorInstantiationError> {
+        if segment_size <= record_size {
+            return Err(ReconstructorInstantiationError::SegmentSizeTooSmall);
+        }
+        if segment_size % record_size != 0 {
+            return Err(ReconstructorInstantiationError::SegmentSizesNotMultipleOfRecordSize);
+        }
+
+        let data_shards = Self::shards_count(segment_size, record_size);
+        let parity_shards = data_shards;
+        let reed_solomon = ReedSolomon::new(data_shards as usize, parity_shards as usize)
+            .expect("ReedSolomon must always be correctly instantiated");
+
+        Ok(Self {
+            record_size,
+            segment_size,
+            reed_solomon,
+            kzg,
+        })
+    }
+
+    /// Returns all the pieces for a segment using given set of pieces of a segment of the archived
+    /// history (any half of all pieces are required to be present, the rest will be recovered
+    /// automatically due to use of erasure coding if needed).
+    pub fn retrieve_pieces(
+        &self,
+        segment_pieces: &[Option<Piece>],
+    ) -> Result<Vec<Piece>, PieceReconstructorError> {
+        // If not all data pieces are available, need to reconstruct data shards using erasure
+        // coding.
+        let mut shards = segment_pieces
+            .iter()
+            .map(|maybe_piece| maybe_piece.as_ref().map(utils::slice_to_arrays))
+            .collect::<Vec<_>>();
+
+        self.reed_solomon
+            .reconstruct(&mut shards)
+            .map_err(PieceReconstructorError::DataShardsReconstruction)?;
+
+        let reconstructed_record_shards = shards
+            .iter()
+            .map(|maybe_shard| {
+                maybe_shard
+                    .as_ref()
+                    .map(|shard| {
+                        let bytes = shard.iter().fold(Vec::new(), |mut acc, shard_part| {
+                            acc.extend_from_slice(shard_part);
+
+                            acc
+                        });
+
+                        let record = &bytes[0..self.record_size as usize];
+
+                        record.to_vec()
+                    })
+                    .expect("Record must be reconstructed here.")
+            })
+            .collect::<Vec<_>>();
+
+        let record_shards_hashes = reconstructed_record_shards
+            .iter()
+            .map(|item| blake2b_256_254_hash(item))
+            .collect::<Vec<_>>();
```

Comment:
Commitments will be much heaver than hash in the future, so I'd like to see a TODO to reuse already present commitments from `segment_pieces` so we don't re-derive what we already have. Also there is a related TODO in archiver about erasure coding of commitments that should probably be replicated here.